AK/Error.h

```diff
@@ -183,14 +183,15 @@ class [[nodiscard]] ErrorOr {
     }
 #endif
 
-    T& value()
-    {
-        return m_value_or_error.template get<T>();
-    }
-    T const& value() const { return m_value_or_error.template get<T>(); }
-
-    ErrorType& error() { return m_value_or_error.template get<ErrorType>(); }
-    ErrorType const& error() const { return m_value_or_error.template get<ErrorType>(); }
+    T& value() & { return m_value_or_error.template get<T>(); }
+    T const& value() const& { return m_value_or_error.template get<T>(); }
+    T&& value() && { return move(m_value_or_error.template get<T>()); }
+    T const&& value() const&& { return move(m_value_or_error.template get<T>()); }
+
+    ErrorType& error() & { return m_value_or_error.template get<ErrorType>(); }
+    ErrorType const& error() const& { return m_value_or_error.template get<ErrorType>(); }
+    ErrorType&& error() && { return move(m_value_or_error.template get<ErrorType>()); }
```

Comment:
This feels like it needs a test. I'm not convinced by this construction that it will actually properly move the Error value out of the m_value_or_error variant if it's an AK::Error.

`Variant<Ts...>` only has two get() overloads, one for T&, and one for T const&.

Your code here is moving the result of `get<ErrorType>()`, which in the case of an ErrorType& is a no-op, and in the case of a `T const&` will always copy.

AK::Error is non-copyable, so I would expect that actually calling this error method on an rvalue/xvalue ErrorOr instance would fail to compile.

In order to get the behavior you want, I believe you would have to spell this as `return move(m_value_or_error).template get<ErrorType>());`, and then add the ref qualified overloads to Variant?

Reply:
perhaps @alimpfard has another view of how this should work?

Reply:
> This feels like it needs a test

Test good, add more tests.

> `Variant<Ts...>` only has two get() overloads, ...

`move(variant.get<T>())` _is_ a move operation (if `Variant<Ts...>& variant`), the semantics of the code here are somewhat silly - the error object is moved-from after this operation, but one part of the error object is technically not moved-from (buuuut that variant index part is a trivial object anyway, so its moved-from state is actually the same as its not-moved-from state).

I would suggest adding `T&& Variant<Ts...>::get<T>() &&` and using `move(m_value_or_error).template get<ErrorType>()` here.